pkg/reconciler/knativeserving/knativeserving.go

```diff
@@ -87,58 +90,71 @@ func (r *Reconciler) FinalizeKind(ctx context.Context, original *servingv1alpha1
 // converge the two.
 func (r *Reconciler) ReconcileKind(ctx context.Context, ks *servingv1alpha1.KnativeServing) pkgreconciler.Event {
 	logger := logging.FromContext(ctx)
+
 	ks.Status.InitializeConditions()
 	ks.Status.ObservedGeneration = ks.Generation
 
 	logger.Infow("Reconciling KnativeServing", "status", ks.Status)
-	stages := []func(context.Context, *mf.Manifest, *servingv1alpha1.KnativeServing) error{
-		r.ensureFinalizerRemoval,
-		r.install,
-		r.checkDeployments,
-		r.deleteObsoleteResources,
-	}
 
-	manifest, err := r.transform(ctx, ks)
+	// Get the target Serving Manifest to be installed
+	targetManifest, err := r.getTargetManifest(ctx, ks)
 	if err != nil {
 		ks.Status.MarkInstallFailed(err.Error())
 		return err
 	}
 
+	// Get the current Manifest, which has been installed.
+	currentManifest, err := r.getCurrentManifest(ctx, ks)
+	if err != nil {
+		return err
+	}
+
+	stages := []func(context.Context, *mf.Manifest, *servingv1alpha1.KnativeServing) error{
+		r.ensureFinalizerRemoval,
+		r.install,
+		r.checkDeployments,
+	}
+
 	for _, stage := range stages {
-		if err := stage(ctx, &manifest, ks); err != nil {
+		if err := stage(ctx, &targetManifest, ks); err != nil {
 			return err
 		}
 	}
+
+	// Remove the resources, that does not exist in the new Serving manifest
+	if err := currentManifest.Filter(mf.None(mf.In(targetManifest))).Delete(); err != nil {
+		return err
+	}
+
 	logger.Infow("Reconcile stages complete", "status", ks.Status)
 	return nil
 }
 
 // Transform the resources
-func (r *Reconciler) transform(ctx context.Context, instance *servingv1alpha1.KnativeServing) (mf.Manifest, error) {
+func (r *Reconciler) transform(ctx context.Context, instance *servingv1alpha1.KnativeServing) ([]mf.Transformer, error) {
 	logger := logging.FromContext(ctx)
 
 	logger.Debug("Transforming manifest")
 
 	platform, err := r.platform.Transformers(r.kubeClientSet, logger)
 	if err != nil {
-		return mf.Manifest{}, err
+		return []mf.Transformer{}, err
 	}
 
 	transformers := common.Transformers(ctx, instance)
 	transformers = append(transformers,
 		ksc.GatewayTransform(instance, logger),
 		ksc.CustomCertsTransform(instance, logger),
 		ksc.HighAvailabilityTransform(instance, logger),
-		ksc.AggregationRuleTransform(r.config.Client))
-	transformers = append(transformers, platform...)
-	return r.config.Transform(transformers...)
+		ksc.AggregationRuleTransform(r.mfClient))
+	return append(transformers, platform...), nil
 }
 
 // Apply the manifest resources
 func (r *Reconciler) install(ctx context.Context, manifest *mf.Manifest, instance *servingv1alpha1.KnativeServing) error {
 	logger := logging.FromContext(ctx)
 	logger.Debug("Installing manifest")
-	return common.Install(manifest, version.ServingVersion, &instance.Status)
+	return common.Install(manifest, instance.Spec.GetVersion(), &instance.Status)
```

Comment:
Can `instance.Spec.GetVersion()` return "" here? Should the operator set it if the user doesn't? Will `common.Install` handle "" predictably?

Reply:
Right, I need to makes sure there is a valid version.